device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java

```diff
@@ -90,8 +90,13 @@
     // Proton-j primitives and wrappers for the device and authentication sessions
     private Connection connection;
     private Reactor reactor;
-    private final Queue<AmqpsSessionHandler> reconnectingDeviceSessionHandlers = new ConcurrentLinkedQueue<>();
-    private final Queue<AmqpsSessionHandler> sessionHandlers = new ConcurrentLinkedQueue<>();
+
+    // keys are device Ids, values are the session handlers associated with that device id
+    private final Map<String, AmqpsSessionHandler> reconnectingDeviceSessionHandlers = new ConcurrentHashMap<>();
```

Comment:
Should you clear this map in `clearLocalState()`?

I fear that when unexpected errors happen while closing a connection that we might run into the original issue again: The timer-task still fires (because it was not cleaned up for some reason) and a session is left over in `reconnectingDeviceSessionHandlers`.

Then messages would be re-queued endlessly just like before.

Reply:
Actually, after taking a second look at this, I can't don that. The ```reconnectingDeviceSessionHandlers``` are used to persist subscription states between connection attempts. For instance, if a multiplexed device is subscribed to twin, runs into a disconnect, and tries to reconnect, this map's saved state of the device's previous session allows for us to implicitly re-subscribe. Clearing this on open/close would make it so that you'd have to re-subscribe to twin and/or methods for all of your clients manually